java/dagger/internal/codegen/BindingGraph.java

```diff
@@ -1090,31 +1088,71 @@ private boolean dependsOnLocalBindingsUncached(Binding binding) {
+              && !getLocalExplicitBindings(keyFactory.unwrapOptional(parentBindings.key()).get())
                   .isEmpty();
         }
+
+        /**
+         * Returns {@code true} if {@code parentBindings} contains a binding for which there is
+         * an explicit binding in this component that was resolved previously in a parent component.
+         */
+        boolean hasLocalContributionBinding(ResolvedBindings parentBindings) {
+          return parentBindings
+                .contributionBindings()
+                .stream()
+                .anyMatch(contributionBinding ->
+                    !getLocalExplicitBindings(parentBindings.key()).isEmpty()
+                    // TO BE REMOVED BEFORE COMMITTING:
+                    //
+                    // Checks that the owning component is not this component.
+                    //
+                    // If the owning component is this component this means the binding that
+                    // requires {@code parentBindings} as a dependency can't be satisfied in the
+                    // parent. This will be reported as a "Cannot be provided" error when the
+                    // graph is validated.
+                    //
+                    // If the owning component isn't this component we have a local contribution
+                    // binding that's a duplicate and will result in a duplicate binding error
+                    // when the graph is validated.
+                    //
+                    // Without this check, GraphValidationTest.bindingUsedOnlyInSubcomponentDependsOnBindingOnlyInSubcomponent()
+                    // fails with the stacktrace below.
+                    //
+                    // java.util.NoSuchElementException
+	                  // at com.google.common.collect.Iterators$1.next(Iterators.java:81)
+	                  // at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:308)
+	                  // at com.google.common.collect.Iterables.getOnlyElement(Iterables.java:263)
+	                  // at dagger.internal.codegen.ResolvedBindings.contributionBinding(ResolvedBindings.java:285)
+	                  // at dagger.internal.codegen.MemberSelect.staticMemberSelect(MemberSelect.java:86)
+	                  // at dagger.internal.codegen.BindingExpression$Factory.forStaticMethod(BindingExpression.java:86)
+	                  // at dagger.internal.codegen.AbstractComponentWriter.createBindingExpression(AbstractComponentWriter.java:419)
+	                  // at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:408)
+	                  // at dagger.internal.codegen.AbstractComponentWriter.createBindingExpressions(AbstractComponentWriter.java:411)
+	                  // at dagger.internal.codegen.AbstractComponentWriter.write(AbstractComponentWriter.java:286)
+                    && parentBindings.owningComponent() != componentDescriptor);
```

Comment:
@netdpb Thanks for the explanation. Unfortunately I don't think the changes you're proposing would translate well to a pull request so I'll leave them to you.

Is there any interest in upstreaming some of the improvements in test case coverage here? Specifically:

1.) The functional test `SubcomponentMembersInjectTest`
2.) `subcomponentBindingConflictsWithParentWhenSubcomponentPerformsMembersInjection()`
3.) `subcomponentGeneratedCorrectlyWhenBothComponentsPerformMembersInjection()`

You may find want to add the following tests back after addressing the problems with `LocalDependencyChecker`:

1. `subcomponentBindingConflictsWithParentWhenBothComponentsPerformMembersInjection()`
2. `subcomponentTransitiveBindingConflictsWithParent()`